### ISSUE TITLE
Don't make aws secret key and token required on AWS backend

### ISSUE THREAD

**Issue by zbintliff** [turbinelabs/rotor#17]
Should fall back to default provider chain. That is task role then instance role, no need to pass sensitive data via env variables.

**Comment by 9len** [comment]
Thanks for the report, @zbintliff. Is this for the ECS or the AWS backend, or both? I see how the SDK running within an ECS task would have access to the right environment variables to do this, but it's not obvious to how the plumbing would work when running within docker on a bare EC2 instance.

**Comment by zbintliff** [comment]
Its actually for both!  The default chain in all AWS supported libraries is as [follows](https://docs.aws.amazon.com/sdk-for-go/v1/developer-guide/configuring-sdk.html#specifying-credentials):

- Use credentials passed explicitly
- Use env variables
- Default to the specified profile or `default` in the config (not valid on ECS or EC2)
- If env is configured for ECS Task, go to metadata endpoint  on every ec2 instance to get credentials for specific container
- Fall back to metadata endpoint to get role for instance profile role.

Where it is set in the go-sdk: https://github.com/aws/aws-sdk-go/blob/f49f6e292a217bb02a127352a379d49bb5106a8d/aws/defaults/defaults.go#L92-L101

This means no need to pass API secret key, or task token. Instead use the AWS native instance profile role and task roles.

**Comment by zbintliff** [comment]
For more information about the meta-data endpoint:
For ecs: https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task-metadata-endpoint.html
For ec2: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-metadata.html

For example if you are on an ec2 instance you can run the following to get your creds:
```
curl 169.254.169.254/latest/meta-data/iam/security-credentials/
```
it will return the name of the instance profile role! Then if you append that to the above curl:

```
curl 169.254.169.254/latest/meta-data/iam/security-credentials/{outputFromAbove}
```
You'll get an output like:
```
{
  "Code" : "Success",
  "LastUpdated" : "redacted",
  "Type" : "redacted",
  "AccessKeyId" : "yyyy",
  "SecretAccessKey" : "xxx",
  "Token" : "xxxxxx",
  "Expiration" : "redacted"
}
```
The sdk will then use that data for the session.

**Comment by 9len** [comment]
Neat. It sounds like on ECS we could just change the flags not to be required and it would all work.

How would this work when running rotor on Docker on a bare AWS instance? Would you have to thread through the `AWS_CONTAINER_CREDENTIALS_FULL_URI` env var? 

What about running in a k8s cluster on AWS instances? In that case, the `AWS_CONTAINER_CREDENTIALS_FULL_URI` var is not readily available. Or is there a metadata endpoint that hands out role based on origin IP?

**Comment by zbintliff** [comment]
Few things to break down there:
>How would this work when running rotor on Docker on a bare AWS instance? Would you have to thread through the AWS_CONTAINER_CREDENTIALS_FULL_URI env var?

For us, when running docker on EC2 not through ECS we just let it use the instance profile role. Since you can still pass the creds via CLI or ENV now its not the biggest deal. if you really worried about security you can even use the .config file in the container.

>What about running in a k8s cluster on AWS instances? In that case, the AWS_CONTAINER_CREDENTIALS_FULL_URI var is not readily available. Or is there a metadata endpoint that hands out role based on origin IP?

Wouldn't rotor in this example then use the k8s backend and not need ec2 credentials?

**Comment by 9len** [comment]
> Wouldn't rotor in this example then use the k8s backend and not need ec2 credentials?

Fair point!

My biggest concern with just marking the credentials as optional will lead to some inscrutable auth messages in the bare-EC2 case. The client only blows up with bad credentials when you actually make a call, so it would be hard to re-wrap errors in a useful way. We can document behavior though.

The other more tactical concern is that we don't have bandwidth to test this change well right now. If you'd like to test (on both ECS and bare EC2) and submit a patch, I'd be happy to merge it.

**Comment by zbintliff** [comment]
Ok, I'll see if I can't get time for it.

**Comment by 9len** [comment]
Thanks! The main changes are:

1. remove the `usage.Required()` wrapper from the [two flags](https://github.com/turbinelabs/rotor/blob/master/plugins/aws/client_fromflags.go#L53-L64)
2. only pass the static credentials [when constructing the session](https://github.com/turbinelabs/rotor/blob/master/plugins/aws/client_fromflags.go#L79-L82) if either of the flags is non-empty

I don't think we have specific test coverage of `aws.clientFromFlags`, so testing will mostly be trying it in ECS and documenting what needs to be done for EC2, probably in the [CLI usage text](https://github.com/turbinelabs/rotor/blob/master/plugins/aws/aws.go#L45-L70) and the [README](https://github.com/turbinelabs/rotor/blob/master/README.md#ec2)

**Comment by zbintliff** [comment]
hey @9len can you elaborate on why the deps of `rotor` (api, cache, etc) are ignored in the `Gopkg.toml`? I understand the version must match across all of them but isn't it safer to pin there then?

**Comment by zbintliff** [comment]
FWIW i just updated the dockerfile to make a test one so i don't need to install all the deps.  usually i use glide or dep but this works.

**Comment by 9len** [comment]
> hey @9len can you elaborate on why the deps of rotor (api, cache, etc) are ignored in the Gopkg.toml? I understand the version must match across all of them but isn't it safer to pin there then?

This is a complication of how we do our deps internally; everything is in a monorepo, and we slice out these public repos. The ignores in `Gopkg.toml` enforce that the monorepo versions of these packages are used. 

If you're on head, it's always safe to just `go get` them, if you're on a specific tagged version, you'd need to mod the `Gopkg.toml` to add those specific versions.

If this becomes a problem for people, we'll ship a separate `Gopkg.toml` with the versions filled in.

**Comment by 9len** [comment]
(we maintain consistent tags across all packages, so if you're using 0.18.0 of rotor, you can safely use 0.18.0 of api, nonstdlib, etc.)

**Comment by zbintliff** [comment]
Took a crack at it in #20 . Didn't do testing or anything yet but wanted to see if you were ok with that approach.

**Comment by 9len** [comment]
fixed by 1ac2787051a3da4f5e8f1249312e3218dac66d46

Thanks @zbintliff !
